3ds2/src/main/java/com/adyen/checkout/adyen3ds2/Adyen3DS2Component.kt

```diff
@@ -0,0 +1,325 @@
+    BaseActionComponent<Adyen3DS2Configuration>(application, configuration), ChallengeStatusReceiver {
+
+    private var mTransaction: Transaction? = null
+    private var mUiCustomization: UiCustomization? = null
+
+    override fun onCleared() {
+        super.onCleared()
+        Logger.d(TAG, "onCleared")
+        if (mTransaction != null) {
+            // We don't save the mTransaction reference if the ViewModel gets cleared.
+            sGotDestroyedWhileChallenging = true
+        }
+    }
+
+    override fun observe(lifecycleOwner: LifecycleOwner, observer: Observer<ActionComponentData>) {
+        super.observe(lifecycleOwner, observer)
+        if (sGotDestroyedWhileChallenging) {
+            // This is OK if the user goes back and starts a new payment.
+            // TODO notify error to activity?
+            Logger.e(TAG, "Lost challenge result reference.")
+        }
+    }
+
+    /**
+     * Set a [UiCustomization] object to be passed to the 3DS2 SDK for customizing the challenge screen.
+     * Needs to be set before handling any action.
+     *
+     * @param uiCustomization The customization object.
+     */
+    fun setUiCustomization(uiCustomization: UiCustomization?) {
+        mUiCustomization = uiCustomization
+    }
+
+    override fun getSupportedActionTypes(): List<String> {
+        return Collections.unmodifiableList(
+            listOf(Threeds2FingerprintAction.ACTION_TYPE, Threeds2ChallengeAction.ACTION_TYPE, Threeds2Action.ACTION_TYPE)
+        )
+    }
+
+    @Throws(ComponentException::class)
+    override fun handleActionInternal(activity: Activity, action: Action) {
+        when (action.type) {
+            Threeds2FingerprintAction.ACTION_TYPE -> {
+                val fingerprintAction = action as Threeds2FingerprintAction
+                if (fingerprintAction.token.isNullOrEmpty()) {
+                    throw ComponentException("Fingerprint token not found.")
+                }
+                identifyShopper(activity, fingerprintAction.token.orEmpty())
+            }
+            Threeds2ChallengeAction.ACTION_TYPE -> {
+                val challengeAction = action as Threeds2ChallengeAction
+                if (challengeAction.token.isNullOrEmpty()) {
+                    throw ComponentException("Challenge token not found.")
+                }
+                challengeShopper(activity, challengeAction.token.orEmpty())
+            }
+            Threeds2Action.ACTION_TYPE -> {
+                val threeds2Action = action as Threeds2Action
+                if (threeds2Action.token.isNullOrEmpty()) {
+                    throw ComponentException("3DS2 token not found.")
+                }
+                if (threeds2Action.subtype == null) {
+                    throw ComponentException("3DS2 Action subtype not found.")
+                }
+                val subtype = parse(threeds2Action.subtype.orEmpty())
```

Comment:
Maybe it's clearer to use `SubType.parse` here and remove the `SubType.Companion.parse` import?